Fix support for multiple GPU

manio opened the pull request:
Discussion: https://github.com/swaywm/wlroots/issues/934

emersion commented on backend/drm/backend.c:
```diff
@@ -51,7 +51,11 @@ static void backend_destroy(struct wlr_backend *backend) {
 	struct wlr_drm_backend *drm = (struct wlr_drm_backend *)backend;
-	return drm->renderer.wlr_rend;
+
+	if (drm->parent)
```
Style: we always use brackets.

Returning the parent's renderer is the Right Thing To Do™.

emersion commented on backend/multi/backend.c:
```diff
@@ -140,15 +140,6 @@ bool wlr_multi_backend_add(struct wlr_backend *_multi,
 		return true;
 	}
 
-	struct wlr_renderer *multi_renderer =
```
I'd really like to keep this.

manio replied:
If I return here (no matter if I return true or false) then the further code for allocating subbackend is not called and this is leading to only support one GPU...

emersion replied:
You can add `backend_renderer != multi_renderer`

emersion commented on render/gles2/renderer.c:
```diff
@@ -25,7 +25,6 @@ static struct wlr_gles2_renderer *gles2_get_renderer(
 static struct wlr_gles2_renderer *gles2_get_renderer_in_context(
 		struct wlr_renderer *wlr_renderer) {
 	struct wlr_gles2_renderer *renderer = gles2_get_renderer(wlr_renderer);
-	assert(wlr_egl_is_current(renderer->egl));
```
We still need to assert this for functions called between `wlr_renderer_begin` and `wlr_renderer_end`. I'd suggest keeping this function as is and adding a new `gles2_get_renderer_switch_context` for textures.

emersion commented on render/gles2/texture.c:
```diff
@@ -25,6 +25,8 @@ static struct wlr_gles2_texture *gles2_get_texture(
 struct wlr_gles2_texture *get_gles2_texture_in_context(
 		struct wlr_texture *wlr_texture) {
 	struct wlr_gles2_texture *texture = gles2_get_texture(wlr_texture);
+	if (!wlr_egl_is_current(texture->egl))
```
This `if` is strange. No need for an assert after switching contexts.

ascent12 commented:
Hmm, this is a smaller change than I was expecting.
There are still lots of things I want to do to make Multi-GPU better (like removing the renderer copy and fixing the code structure), which is probably why I was being damn slow/lazy when I was looking into this myself instead of looking at the core issue.

I'll get around to testing a few of the various GPU configurations I have access to this weekend.

To anyone else trying this out, pay attention to:
- Hotplugging
- Using non-native resolutions and rotations/reflections
- Whether things like hardware cursors and gamma control work
- The overall performance (stuttering, etc.)
- Different configurations of which GPU is primary (e.g. run with WLR_DRM_DEVICES=/dev/dri/card0:/dev/dri/card1 then WLR_DRM_DEVICES=/dev/dri/card1:/dev/dri/card0)
- Whether this works on all of rootston, sway, and any other wlroots-based compositor
- Whether tri or quad-GPU systems work well (I only have access to motherboards with enough PCI-e space for 2 GPUs)
- Whether it works cross driver (i915 + nouveau, amdgpu + radeon, etc.)

manio commented:
I've updated the code one more time. I was switching context when I did not have to and this was leading to problems (non visual, but in the log, like):
```[GLES2] GL_INVALID_OPERATION ...```

This is now fixed. All @emersion comments are now took into account.

There are still some problems left:
- Screen mirror + rotation
- Hardware cursor on secondary display

But I'd like to have it covered in another issues/pull requests (step by step).

manio commented:
Screen rotation fixed :)

emersion commented on render/gles2/renderer.c:
```diff
@@ -323,7 +323,7 @@ static struct wlr_texture *gles2_texture_from_dmabuf(
 		struct wl_display *wl_display) {
 	struct wlr_gles2_renderer *renderer =
-		gles2_get_renderer_in_context(wlr_renderer);
+		gles2_get_renderer(wlr_renderer);
```
Hmm. We should probably switch context here.

manio replied:
Tried one more time after your comment. If I revert this change I have the assert on startup.

manio replied:
Ah - I probably took you wrong. You are telling about switching context and I was thinking about reverting this change to original function (which is asserting and not switching context, as this was your point regarding rendering functions)... Yeah - I can switch context here as well.

emersion commented on render/gles2/renderer.c:
```diff
@@ -227,7 +227,7 @@ static bool gles2_resource_is_wl_drm_buffer(struct wlr_renderer *wlr_renderer,
 		struct wl_resource *buffer, int *width, int *height) {
 	struct wlr_gles2_renderer *renderer =
-		gles2_get_renderer_in_context(wlr_renderer);
+		gles2_get_renderer(wlr_renderer);
```
Not sure if we should switch context here. @ascent12, thoughts?

manio replied:
I think it should be as it is now. I've also retested reverting this change - after some working in sway and opening some XWayland apps the assert is raising.

emersion commented on render/gles2/renderer.c:
```diff
@@ -116,7 +116,7 @@ static bool gles2_render_texture_with_matrix(struct wlr_renderer *wlr_renderer,
 		gles2_get_renderer_in_context(wlr_renderer);
 	struct wlr_gles2_texture *texture =
-		get_gles2_texture_in_context(wlr_texture);
+		gles2_get_texture(wlr_texture);
```
Can you try adding back this assertion? I'd like to assert that the texture has been created in the same context as the renderer.

ascent12 commented:
I've pushed the fix for the black hardware cursor on the secondary GPU.

manio commented:
@ascent12 
Thank you for your work! Confirming - the cursor is now working, yay! :)

If you don't have nothing against I think it's now in good shape to merge this PR.

n3rdopolis commented:
I'm having a little trouble starting this on my qemu VM with multiple GPUs.
The relevant part of my QEMU command is
```
qemu-system-x86_64 -m 2047 -vga none -enable-kvm -device VGA,id=video0  -device secondary-vga,id=video1 -spice port=5930,disable-ticketing -cpu host -cdrom PATHTOMYISO.iso
```
to get two GPUs. and then to see the VM with the two GPUs:
```
remote-viewer spice://127.0.0.1:5930 --spice-preferred-compression=off
```
rootston and sway both abort with
```
sway: ../backend/drm/renderer.c:238: copy_drm_surface_mgpu: Assertation 'tex' failed.
```

Sway can start when there is one GPU, Let me know if you need more info

SirCmpwn commented:
Let's take a look at qemu separately. Can you open a ticket?

Thanks for the patch, @manio!
